homeassistant/util/unit_system.py

```diff
@@ -167,6 +167,37 @@ def as_dict(self) -> dict:
             SPEED: self.speed_unit
         }
 
+    def convert(self, state, unit_of_measure) -> Tuple[float, str]:
+        """Generic conversion method."""
+        try:
+            if (unit_of_measure in (TEMP_CELSIUS, TEMP_FAHRENHEIT) and
+                    unit_of_measure != self.temperature_unit):
+                prec = len(state) - state.index('.') - 1 if '.' in state else 0
```

Comment:
These three  codeblocks that start with this line look very similar. I think you should turn them around so that you first check what method to use for conversion and store that method in a new variable. Then outside of the if/elif blocks you do the conversion and follow up with rounding, cause that part will be the same for all three blocks.